### ISSUE TITLE
How to write codes about `argtypes` , `memsync` and `restype` for `EnumWindows` ?

### ISSUE THREAD

**Issue by yssource** [pleiszenburg/zugbruecke#23]
@s-m-e Thanks a lot for your project.

How to write codes about `argtypes` , `memsync` and `restype` for `EnumWindows` ?
Could you help me and give me a example?

```c++
BOOL WINAPI EnumWindows(
  _In_ WNDENUMPROC lpEnumFunc,
  _In_ LPARAM      lParam
);
```

**Comment by s-m-e** [comment]
@yssource You're welcome :)

I am not entirely sure about your usecase for `EnumWindows` (a Windows API for enumerating all open Windows on the user's screen?), but I will look into the definition. It is likely going to enumerate all open windows of processes running on top of the current Wine server if you do try to call it on a Unix system.

**Comment by yssource** [comment]
- [My original snippet](#sec-1)
- [My trial](#sec-2)

Here is snippet of my codes which work well under windows with ctypes.

And now, I try to modify it in order to work with zugbruecke.

# My original snippet<a id="sec-1"></a>

```python
import zugbruecke as ctypes
api = ctypes.windll.user32
buff = ctypes.create_unicode_buffer(32)


def find(keyword='My-Client):
    """ Enumerate all the registered clients """
    clients = set()

    # @ctypes.WINFUNCTYPE(ctypes.c_long, ctypes.c_void_p, ctypes.c_wchar_p)
    @ctypes.CFUNCTYPE(ctypes.c_long, ctypes.c_void_p, ctypes.c_wchar_p)
    def check(hwnd, keyword):
        """ callback function, Selection of key words in the title bar """
        if api.IsWindowVisible(hwnd) and api.GetWindowTextW(hwnd, buff, 32) > 6 and keyword in buff.value:
            clients.add(hwnd)
        return 1

    api.EnumWindows(check, keyword)

    return clients


```
-   I noteice that `WINFUNCTYPE` is not supported in zugbruecke, so I change it to `CFUNCTYPE`

![screenshot from 2018-03-10 09-16-33](https://user-images.githubusercontent.com/3756334/37237501-ef737d18-244e-11e8-88f7-36d7528e443d.png)

# My trial<a id="sec-2"></a>

```python
# ... some other codes above

api.EnumWindows.argtypes = (ctypes.POINTER(ctypes.c_void_p), ctypes.c_wchar_p)
api.EnumWindows.memsync = [
        {
                'p': [0],
                'l': [1],
                '_t': ctypes.c_void_p
        }
]
api.EnumWindows.restype = ctypes.c_bool
api.EnumWindows(check, keyword)

return clients

```
![screenshot from 2018-03-10 10-33-03](https://user-images.githubusercontent.com/3756334/37237507-f9171a5a-244e-11e8-8828-472dd747f686.png)

**Comment by s-m-e** [comment]
@yssource I see where you problem is. `zugbruecke` does not support [call back functions](https://en.wikipedia.org/wiki/Callback_(computer_programming)) as argument types (yet). As you have already discovered, *both* types of call back function pointers, `CFUNCTYPE` and `WINFUNCTYPE`, are unsupported at this point. In your snipped / trial, you are essentially trying to pass a pointer to a (call back) function as a parameter into your call of `EnumWindows`. `zugbruecke` then tries to serialize the data the pointer is pointing to (which is actually a function) and, because it does not know what a function pointer is, crashes subsequently. 

Long story short: I have been asked to add support for function pointers before. Personally, I never had a use case so I never implemented it. With the current code base of  `zugbruecke`, it is not overly complicated to add this feature, but I do not know when I will find the time for this in the foreseeable future. If you want to take a stab at this in the meantime, I'd be more than happy to assist you. I am sorry I can not help you any further right now.

**Comment by yssource** [comment]
Thanks all the same.
I'd like take more time to learn your project and try to find a solution of the callback.

**Comment by yssource** [comment]
Personally I think it is not possible to run `WINFUNCTYPE` on client-end, since it is only able to run `WINFUNCTYPE` under win32. Isn't it? 
And, do you think that my snipped should be run under wine?

**Comment by s-m-e** [comment]
@yssource I have written a small (not functional) code skeleton which I am going upload with a few comments.

Concept: Have a look at how I implemented the `WinDLL` method (in `_wrapper_.py`). In `ctypes`, `WinDLL` is only available on Windows. In `zugbruecke`, I have implemented my own version of `WinDLL` on the Unix Python side, which accesses an instance ("session") of the `session_client_class` (see `session_client.py`). The session client then accesses the Wine Python server, where it talks to the session server (see `session_server.py`). The session server can access a real Windows version of `ctypes`, so it can use the actual `ctypes` implementation of `WinDLL`. 

How function pointers are implemented in `ctypes`: Now have a look at how `WINFUNCTYPE` and `CFUNCTYPE` are implemented in the original `ctypes`. Both are methods, which create new classes "on the fly" for every type of function pointer. The classes then (a) serve as argument definitions and (b) can be used as decorators on Python methods. The decoration turns the Python methods into a `ctypes` function pointer ("call back object"). The latter can be passed into a `ctypes` function call as a parameter. 

How they can be implemented in `zugbruecke`: First, the `WINFUNCTYPE` function from `ctypes` is not available on non-Windows system simply because of an `if` statement in `ctypes` checking the operating system. Otherwise, it is virtually identical to the `CFUNCTYPE` function except for one altered flag. You can copy & paste the code of `WINFUNCTYPE` into your own script and it will run on Linux. (You can not call into a DLL like that, but you can generate a function pointer with the right specs.) So at this stage, you have both `CFUNCTYPE` and `WINFUNCTYPE` available. The next step is to parse this stuff so you can send the specifications to the Wine Python server. I suggest to rename the `CFUNCTYPE` from `ctypes` on the Unix Python client, like e.g. `from ctypes import CFUNCTYPE as __CFUNCTYPE__` . You also copy and paste the original `WINFUNCTYPE` code into `zugbruecke` and rename it into e.g. `__WINFUNCTYPE__`. You then implement your own versions of `CFUNCTYPE` and `WINFUNCTYPE`, which are wrapped around `__CFUNCTYPE__` and `__WINFUNCTYPE__`. Now *your* Unix Python client `CFUNCTYPE` and `WINFUNCTYPE` functions you can "play" with the following things: A handle on the Python call back function, the definition of its arguments in `ctypes` terms and its return type definition in `ctypes` terms. Both sets of definitions can be parsed by the code found in `args_definition.py` and shipped to the server, where they can be reconstructed into `ctypes` definitions (also by the code found in `args_definition.py`). Now, on the Unix Python client side, you have to add another remote procedure call (RPC) server. You'll find all the code you need in `rpc.py`. `zugbruecke` is already starting two RPC servers, one on the Unix Python client side and one on the Wine Python server side. The RPC server on the Unix Python client side accepts logging messages from the Wine Python server (this is how the logs in your screenshots were produced). The RPC server on the Wine Python server side exposes the DLL routines to the Unix Python client, so your Unix Python client can execute DLL routines on the Wine Python server. I'd say you simply add a third RPC server to the mix, on the Unix Python client side, which exposes the call back functions to the Wine Python server. (I know, I am heavily mixing up the terms "server" and "client" here ... it's confusing.) Now you have all you need: On the Wine Python server side, you have the argument and return type definitions available, you have the actual `ctypes` implementations of `CFUNCTYPE` and `WINFUNCTYPE` available, and you have the call backs from the Unix Python client through an RPC client available. All you have to do now is to build a wrapper routine on the Wine Python server, which parses the actual arguments and ships them to the call back function on the Unix Python client side where they have to be rebuild into `ctypes` arguments (all necessary code is found in `arg_contents.py`).

Well, the above text is simplifying things a bit, but most of the required code and infrastructure is already implemented in `zugbruecke`. You'd basically "only" have to build a "reversed" or "mirrored" call mechanism, from the Wine Python server back to the Unix Python client. The mechanics are almost identical to the mechanics of a regular call into a DLL from the Unix Python client to the Wine Python server.

**Comment by s-m-e** [comment]
@yssource I turned my non-working code skeleton into a **highly experimental working patch**. It was easier than I expected, so I just finished it. There is now (limited) support for function pointers, both types (`CFUNCTYPE` and `WINFUNCTYPE`). I merged the patch into the [development branch](/pleiszenburg/zugbruecke/tree/develop). [This example](https://github.com/pleiszenburg/zugbruecke/blob/develop/examples/test_callback.py) is now working with `zugbruecke`. Please test this new feature and let me if it works for your application / use-case.

There is significant limitation: The DLL can pass pointers into the callback function. My code does support that. If a DLL passes a pointer pointing to memory, which had previously been allocated by the Unix Client, the callback function on the Unix Client will not receive the same pointer. It will instead receive a new one, pointing to a copy of the respective memory. This is (a) slow and (b) breaks a number of relevant use-cases for callback functions. Fixing this requires a significant re-write of the entire code for handling memory and pointers across all of `zugbruecke`.

**Comment by yssource** [comment]
Your callback example works well for me.

But if the callback call another subroutine, it will be stuck at `UNIX o: [dll-client] ... unknown, registering`.

For example, `check` call another `IsWindowVisible` in my codes.

I think the problem is it was supposed that `IsWindowVisible` should has a different `hash_id_register_routine` for itself, but actually the `[[https://github.com/pleiszenburg/zugbruecke/blob/develop/src/zugbruecke/core/dll_client.py#L66][self.__register_routine_on_server__]]` is only initialized once. And subroutine shares only one hasd_id with its parent's.

**Comment by s-m-e** [comment]
@yssource `[hash_id]_register_routine` exists only one single time per DLL. One DLL is represented by one instance of the `dll_client_class` on the Unix side and by one instance of the `dll_server_class` on the Wine side. I am simply linking those two instances together. I.e. the `hash_id` uniquely represents one single DLL. `IsWindowVisible` is a routine and has (almost) nothing to do with the `hash_id` of its parent DLL.

Let's look at your code. I modified it a bit for easier debugging and added all missing `argtypes` and `restype` definitions:
```

from sys import platform

if any([platform.startswith(os_name) for os_name in ['linux', 'darwin', 'freebsd']]):

	f = open('.zugbruecke.json', 'w')
	f.write('{"log_level": 10}')
	f.close()

	import zugbruecke as ctypes
	from zugbruecke import wintypes

elif platform.startswith('win'):

	import ctypes
	from ctypes import wintypes

else:

	raise # TODO unsupported platform


api = ctypes.windll.user32
buff = ctypes.create_unicode_buffer(32)

check_ptr_type = ctypes.WINFUNCTYPE(ctypes.c_long, wintypes.HWND, ctypes.c_wchar_p)

IsWindowVisible = api.IsWindowVisible
IsWindowVisible.argtypes = (wintypes.HWND,)
IsWindowVisible.restype = wintypes.BOOL

GetWindowTextW = api.GetWindowTextW
GetWindowTextW.argtypes = (wintypes.HWND, ctypes.c_wchar_p, ctypes.c_int)
GetWindowTextW.restype = ctypes.c_int
GetWindowTextW.memsync = [
	{
		'p': [1],
		'l': [2],
		'_t': wintypes.WCHAR
		}
	]

EnumWindows = api.EnumWindows
EnumWindows.argtypes = (check_ptr_type, wintypes.LPARAM)
EnumWindows.restype = wintypes.BOOL


def find(keyword = 'My-Client'):
	""" Enumerate all the registered clients """

	clients = set()

	@check_ptr_type
	def check(hwnd, keyword):
		""" callback function, Selection of key words in the title bar """

		print('Check!')
		_vis = IsWindowVisible(hwnd) # BUG: Pointer is "translated" into new pointer and not kept in sync
		print(_vis)
		_text = GetWindowTextW(hwnd, buff, 32)
		print((_text, buff.value))

		if _vis and _text > 6 and keyword in buff.value:
			clients.add(hwnd)

		return 1

	EnumWindows(check, 0) # keyword is a string, but EnumWindows expects an integer/long as second parameter

	return clients


if __name__ == '__main__':

	print(find())
```

The problem is triggered by `_vis = IsWindowVisible(hwnd)` (it just hangs). This is exactly what I mean by "limited support" for callback functions, I am afraid. `hwnd` is a pointer, which is send back and forth between the Unix and the Wine side. Unfortunately, it is not kept in sync properly. It is allocated on the Wine side, sent to the Unix side and sent back to the Wine side. When it again reaches the Wine side, it is not the same pointer anymore. I suppose this can be fixed, but it is rather complicated thing to do.

**Comment by stuaxo** [comment]
OT:  @yssource I had to write something like this to implement a program to [screen capture windows](https://github.com/stuaxo/capture_window/blob/master/capture_window/capture_window.cpp), but in C++ (which I don't really know), it would be interesting if you get this working as I'd really like to implement the same code in python.

Unfortunately, while it worked in Windows, I couldn't get it to capture anything but a black image in Wine, but not sure if that was my code or a Wine bug... definitely would be quicker to iterate in python though.

**Comment by yssource** [comment]
@stuaxo 
I'm sorry. I didn't take a close look into how the callback is working here.
My Host is Linux, and the guest is wine.
I think it would be better to capture the windows under the Linux side, since there are many ways to do it with xdg tools, such as xwininfo, xdotool, and so on.

**Comment by s-m-e** [comment]
This issue is now basically equal to #25. It's a symmetry (or lack thereof) issue. I attempted to produce a graphic depiction of this issue:

![issue25](https://user-images.githubusercontent.com/16851110/71685974-26f49c80-2d9a-11ea-8e4d-28aeeb8f6d07.png)

I hope this makes sense. I am working on a re-write of the memory management. It can not be solved with a "fix" - it required rewriting most of the serialization / de-serialization code.

**Comment by s-m-e** [comment]
Tracking this in #25, probably for v0.2.0.
